src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/BlockMerkleRootRule.cs

```diff
@@ -70,63 +73,60 @@ public static uint256 BlockMerkleRoot(Block block, out bool mutated)
         public static uint256 ComputeMerkleRoot(List<uint256> leaves, out bool mutated)
         {
-            var branch = new List<uint256>();
-
             mutated = false;
-            if (leaves.Count == 0)
-                return uint256.Zero;
+            if (leaves.Count == 0) return uint256.Zero;
 
-            // count is the number of leaves processed so far.
-            uint count = 0;
+            var branch = new List<uint256>();
 
-            // inner is an array of eagerly computed subtree hashes, indexed by tree
+            // subTreeHashes is an array of eagerly computed subtree hashes, indexed by tree
             // level (0 being the leaves).
-            // For example, when count is 25 (11001 in binary), inner[4] is the hash of
-            // the first 16 leaves, inner[3] of the next 8 leaves, and inner[0] equal to
-            // the last leaf. The other inner entries are undefined.
-            var inner = new uint256[32];
+            // For example, when count is 25 (11001 in binary), subTreeHashes[4] is the hash of
+            // the first 16 leaves, subTreeHashes[3] of the next 8 leaves, and subTreeHashes[0] equal to
+            // the last leaf. The other subTreeHashes entries are undefined.
+            var subTreeHashes = new uint256[32];
 
-            for (int i = 0; i < inner.Length; i++)
-                inner[i] = uint256.Zero;
+            for (int i = 0; i < subTreeHashes.Length; i++)
+                subTreeHashes[i] = uint256.Zero;
 
             // Which position in inner is a hash that depends on the matching leaf.
             int matchLevel = -1;
+            uint processedLeavesCount = 0;
 
-            // First process all leaves into 'inner' values.
-            while (count < leaves.Count)
+            // First process all leaves into subTreeHashes values.
+            while (processedLeavesCount < leaves.Count)
             {
-                uint256 h = leaves[(int)count];
+                uint256 currentLeaveHash = leaves[(int)processedLeavesCount];
                 bool match = false;
-                count++;
+                processedLeavesCount++;
                 int level;
 
-                // For each of the lower bits in count that are 0, do 1 step. Each
-                // corresponds to an inner value that existed before processing the
+                // For each of the lower bits in processedLeavesCount that are 0, do 1 step. Each
+                // corresponds to an subTreeHash value that existed before processing the
                 // current leaf, and each needs a hash to combine it.
-                for (level = 0; (count & (((uint)1) << level)) == 0; level++)
+                // The following loop could also be read like this:
+                // for (level = 0; (processedLeavesCount % ((2^level) -1)) == 0; level++)
+                // but bitwise operations are faster    
```

Comment:
This is wrong I thought I explained that please remove it

Reply:
Ok I would like a proper explanation on this logic @monsieurleberre please understand exactly the low level with apro and then describe it in a comment inline (inside the method).

Reply:
The original comment seems to be very much sufficient

Reply:
@dangershony For more context please see discussion starting here https://github.com/stratisproject/StratisBitcoinFullNode/pull/1353#discussion_r184953228

Reply:
OK I removed it as it is not strictly equivalent you are right, it still helped me a lot to understand the loop though to get a rough understanding of this condition / loop

Reply:
Thank you. But I think you're still misunderstanding. It is not that it is not strictly equivalent, it is completely different. So if what you're saying is true and it helped you understand, then you understand it wrongly.

Take a look
```
 for (level = 0; (count & (((uint)1) << level)) == 0; level++)
```

```
for (level = 0; (processedLeavesCount % ((2^level) -1)) == 0; level++) 
```

So let's say our `count` is `1011000 (bin) = 0x58 = 88 (dec)`, what do we have? 

1st iteration: `level = 0`
original: `(0x58 & (1 << 0)) == 0` is TRUE -> loop is taken
your code: `(88 % (1 - 1)) == 0` crashes

2nd iteration: `level = 1`
original: `(0x58 & (1 << 1)) == 0` is TRUE -> loop is taken
your code: `(88 % (2 - 1)) == 0` is TRUE (by accident because modulo 1 is always 0 for everything, so it is not possible for your code to return FALSE in second iteration)

3nd iteration: `level = 2`
original: `(0x58 & (1 << 2)) == 0` is TRUE -> loop is taken
your code: `(88 % (4 - 1)) == 0` is FALSE -> loop is NOT taken and would end

4nd iteration: `level = 3`
original: `(0x58 & (1 << 3)) == 0` is TRUE -> loop NOT taken and ends
your code: `(88 % (8 - 1)) == 0` is FALSE -> loop is NOT taken and would end


now if you try number `1000010`, your second iteration would do something else than original and you can literally find number that behaves differently on every level

there is NO relation between your code and original, it is completely wrong